File: backend/app/application/services/product_service.py
from typing import List, Optional, Dict, Any
from sqlalchemy.ext.asyncio import AsyncSession

from app.infrastructure.persistence.product_repository import ProductRepository
from app.schemas.product import (
    CategoryCreate, CategoryUpdate, Category,
    ProductCreate, ProductUpdate, Product,
    ProductSKUCreate, ProductSKUUpdate, ProductSKU,
    ProductPriceCreate, ProductPriceUpdate, ProductPrice,
    ProductAttributeCreate, ProductAttribute
)


class ProductService:
    """商品管理服务层"""
    
    def __init__(self, db: AsyncSession):
        self.repository = ProductRepository(db)
    
    # 分类相关服务
    async def create_category(self, category_data: CategoryCreate) -> Category:
        """创建分类"""
        category = await self.repository.create_category(
            name=category_data.name,
            parent_id=category_data.parent_id,
            level=category_data.level,
            sort=category_data.sort
        )
        return Category.model_validate(category)
    
    async def get_category(self, category_id: int) -> Optional[Category]:
        """获取分类详情"""
        category = await self.repository.get_category(category_id)
        if not category:
            return None
        return Category.model_validate(category)
    
    async def get_categories(self, parent_id: Optional[int] = None, is_active: Optional[bool] = None) -> List[Category]:
        """获取分类列表"""
        categories = await self.repository.get_categories(parent_id, is_active)
        return [Category.model_validate(category) for category in categories]
    
    async def update_category(self, category_id: int, category_data: CategoryUpdate) -> Optional[Category]:
        """更新分类"""
        update_data = category_data.model_dump(exclude_unset=True)
        category = await self.repository.update_category(category_id, **update_data)
        if not category:
            return None
        return Category.model_validate(category)
    
    async def delete_category(self, category_id: int) -> bool:
        """删除分类"""
        return await self.repository.delete_category(category_id)
    
    # 商品相关服务
    async def create_product(self, product_data: ProductCreate) -> Product:
        """创建商品"""
        product = await self.repository.create_product(
            name=product_data.name,
            category_id=product_data.category_id,
            description=product_data.description,
            brand=product_data.brand,
            model=product_data.model,
            is_active=product_data.is_active
        )
        return Product.model_validate(product)
    
    async def get_product(self, product_id: int) -> Optional[Product]:
        """获取商品详情"""
        product = await self.repository.get_product(product_id)
        if not product:
            return None
        return Product.model_validate(product)
    
    async def get_products(self, category_id: Optional[int] = None, is_active: Optional[bool] = None, skip: int = 0, limit: int = 100) -> List[Product]:
        """获取商品列表"""
        products = await self.repository.get_products(category_id, is_active, skip, limit)
        return [Product.model_validate(product) for product in products]
    
    async def update_product(self, product_id: int, product_data: ProductUpdate) -> Optional[Product]:
        """更新商品"""
        update_data = product_data.model_dump(exclude_unset=True)
        product = await self.repository.update_product(product_id, **update_data)
        if not product:
            return None
        return Product.model_validate(product)
    
    async def delete_product(self, product_id: int) -> bool:
        """删除商品"""
        return await self.repository.delete_product(product_id)
    
    # SKU相关服务
    async def create_sku(self, product_id: int, sku_data: ProductSKUCreate) -> ProductSKU:
        """创建SKU"""
        sku = await self.repository.create_sku(
            product_id=product_id,
            sku_code=sku_data.sku_code,
            attributes=sku_data.attributes,
            stock=sku_data.stock
        )
        return ProductSKU.model_validate(sku)
    
    async def get_sku(self, sku_id: int) -> Optional[ProductSKU]:
        """获取SKU详情"""
        sku = await self.repository.get_sku(sku_id)
        if not sku:
            return None
        return ProductSKU.model_validate(sku)
    
    async def get_skus_by_product(self, product_id: int) -> List[ProductSKU]:
        """获取商品的所有SKU"""
        skus = await self.repository.get_skus_by_product(product_id)
        return [ProductSKU.model_validate(sku) for sku in skus]
    
    async def update_sku(self, sku_id: int, sku_data: ProductSKUUpdate) -> Optional[ProductSKU]:
        """更新SKU"""
        update_data = sku_data.model_dump(exclude_unset=True)
        sku = await self.repository.update_sku(sku_id, **update_data)
        if not sku:
            return None
        return ProductSKU.model_validate(sku)
    
    async def delete_sku(self, sku_id: int) -> bool:
        """删除SKU"""
        return await self.repository.delete_sku(sku_id)
    
    # 价格相关服务
    async def create_price(self, sku_id: int, price_data: ProductPriceCreate) -> ProductPrice:
        """创建价格"""
        price = await self.repository.create_price(
            sku_id=sku_id,
            price_type=price_data.price_type,
            price=price_data.price,
            min_quantity=price_data.min_quantity,
            max_quantity=price_data.max_quantity
        )
        return ProductPrice.model_validate(price)
    
    async def get_prices_by_sku(self, sku_id: int) -> List[ProductPrice]:
        """获取SKU的所有价格"""
        prices = await self.repository.get_prices_by_sku(sku_id)
        return [ProductPrice.model_validate(price) for price in prices]
    
    async def update_price(self, price_id: int, price_data: ProductPriceUpdate) -> Optional[ProductPrice]:
        """更新价格"""
        update_data = price_data.model_dump(exclude_unset=True)
        price = await self.repository.update_price(price_id, **update_data)
        if not price:
            return None
        return ProductPrice.model_validate(price)
    
    async def delete_price(self, price_id: int) -> bool:
        """删除价格"""
        return await self.repository.delete_price(price_id)
    
    # 属性相关服务
    async def create_attribute(self, product_id: int, attribute_data: ProductAttributeCreate) -> ProductAttribute:
        """创建商品属性"""
        attribute = await self.repository.create_attribute(
            product_id=product_id,
            name=attribute_data.name,
            value=attribute_data.value
        )
        return ProductAttribute.model_validate(attribute)
    
    async def get_attributes_by_product(self, product_id: int) -> List[ProductAttribute]:
        """获取商品的所有属性"""
        attributes = await self.repository.get_attributes_by_product(product_id)
        return [ProductAttribute.model_validate(attribute) for attribute in attributes]
    
    async def delete_attribute(self, attribute_id: int) -> bool:
        """删除商品属性"""
        return await self.repository.delete_attribute(attribute_id)

File: frontend/admin/src/services/productService.ts
import api from './api';
import { ApiResponse, Product, Sku, Category } from '../types';

interface ProductCreateData {
  name: string;
  category_id: number;
  description?: string;
  brand?: string;
  model?: string;
  is_active: boolean;
  image?: string;
}

interface SkuCreateData {
  product_id: string;
  sku_code: string;
  name: string;
  stock: number;
  cost_price: number;
  weight?: number;
  volume?: number;
  attributes?: any;
  is_active: boolean;
}

interface PriceCreateData {
  sku_id: string;
  price_type: string;
  price: number;
  min_quantity?: number;
  max_quantity?: number;
  is_active?: boolean;
}

const productService = {
  // 获取商品列表
  getProducts: async (params: Record<string, any> = {}): Promise<ApiResponse<Product[]>> => {
    try {
      const response = await api.get('/v1/products', { params });
      return response.data;
    } catch (error) {
      console.error('获取商品列表失败:', error);
      throw error;
    }
  },

  // 获取商品详情
  getProduct: async (id: string): Promise<ApiResponse<Product>> => {
    try {
      const response = await api.get(`/v1/products/${id}`);
      return response.data;
    } catch (error) {
      console.error(`获取商品详情失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 创建商品
  createProduct: async (productData: ProductCreateData): Promise<ApiResponse<Product>> => {
    try {
      const response = await api.post('/v1/products', productData);
      return response.data;
    } catch (error) {
      console.error('创建商品失败:', error);
      throw error;
    }
  },

  // 更新商品
  updateProduct: async (id: string, productData: ProductCreateData): Promise<ApiResponse<Product>> => {
    try {
      const response = await api.put(`/v1/products/${id}`, productData);
      return response.data;
    } catch (error) {
      console.error(`更新商品失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 删除商品
  deleteProduct: async (id: string): Promise<ApiResponse<void>> => {
    try {
      const response = await api.delete(`/v1/products/${id}`);
      return response.data;
    } catch (error) {
      console.error(`删除商品失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 获取商品分类列表
  getCategories: async (): Promise<ApiResponse<Category[]>> => {
    try {
      const response = await api.get('/v1/products/categories');
      return response.data;
    } catch (error) {
      console.error('获取商品分类失败:', error);
      throw error;
    }
  },

  // 创建分类
  createCategory: async (categoryData: any): Promise<ApiResponse<Category>> => {
    try {
      const response = await api.post('/v1/products/categories', categoryData);
      return response.data;
    } catch (error) {
      console.error('创建分类失败:', error);
      throw error;
    }
  },

  // 更新分类
  updateCategory: async (id: string, categoryData: any): Promise<ApiResponse<Category>> => {
    try {
      const response = await api.put(`/v1/products/categories/${id}`, categoryData);
      return response.data;
    } catch (error) {
      console.error('更新分类失败:', error);
      throw error;
    }
  },

  // 删除分类
  deleteCategory: async (id: string): Promise<ApiResponse<void>> => {
    try {
      const response = await api.delete(`/v1/products/categories/${id}`);
      return response.data;
    } catch (error) {
      console.error('删除分类失败:', error);
      throw error;
    }
  },

  // 获取SKU列表
  getSkus: async (productId: string): Promise<ApiResponse<Sku[]>> => {
    try {
      const response = await api.get(`/v1/products/${productId}/skus`);
      return response.data;
    } catch (error) {
      console.error(`获取SKU列表失败 (商品ID: ${productId}):`, error);
      throw error;
    }
  },

  // 创建SKU
  createSku: async (skuData: SkuCreateData): Promise<ApiResponse<Sku>> => {
    try {
      // 从skuData中移除product_id字段，因为它是通过URL路径传递的
      const { product_id, ...skuDataWithoutProductId } = skuData;
      const response = await api.post(`/v1/products/${product_id}/skus`, skuDataWithoutProductId);
      return response.data;
    } catch (error) {
      console.error('创建SKU失败:', error);
      throw error;
    }
  },

  // 更新SKU
  updateSku: async (id: string, skuData: SkuCreateData): Promise<ApiResponse<Sku>> => {
    try {
      // 从skuData中移除product_id字段，因为更新SKU时不需要它
      const { product_id, ...skuDataWithoutProductId } = skuData;
      const response = await api.put(`/v1/products/skus/${id}`, skuDataWithoutProductId);
      return response.data;
    } catch (error) {
      console.error(`更新SKU失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 删除SKU
  deleteSku: async (id: string): Promise<ApiResponse<void>> => {
    try {
      const response = await api.delete(`/v1/products/skus/${id}`);
      return response.data;
    } catch (error) {
      console.error(`删除SKU失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 获取价格列表
  getPrices: async (skuId: string): Promise<ApiResponse<any[]>> => {
    try {
      const response = await api.get(`/v1/products/skus/${skuId}/prices`);
      return response.data;
    } catch (error) {
      console.error(`获取价格列表失败 (SKU ID: ${skuId}):`, error);
      throw error;
    }
  },

  // 创建价格
  createPrice: async (priceData: PriceCreateData): Promise<ApiResponse<any>> => {
    try {
      const response = await api.post(`/v1/products/skus/${priceData.sku_id}/prices`, priceData);
      return response.data;
    } catch (error) {
      console.error('创建价格失败:', error);
      throw error;
    }
  },

  // 更新价格
  updatePrice: async (id: string, priceData: PriceCreateData): Promise<ApiResponse<any>> => {
    try {
      const response = await api.put(`/v1/products/prices/${id}`, priceData);
      return response.data;
    } catch (error) {
      console.error(`更新价格失败 (ID: ${id}):`, error);
      throw error;
    }
  },

  // 删除价格
  deletePrice: async (id: string): Promise<ApiResponse<void>> => {
    try {
      const response = await api.delete(`/v1/products/prices/${id}`);
      return response.data;
    } catch (error) {
      console.error(`删除价格失败 (ID: ${id}):`, error);
      throw error;
    }
  },
};

export default productService;